bee-crypto/src/ternary/bigint/i384/mod.rs

```diff
@@ -0,0 +1,893 @@
+// Copyright 2020 IOTA Stiftung
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+// an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and limitations under the License.
+
+//! This module contains signed integers encoded by 384 bits.
+
+mod constants;
+
+pub use constants::{
+    BE_U32_0, BE_U32_1, BE_U32_2, BE_U32_MAX, BE_U32_MIN, BE_U32_NEG_1, BE_U32_NEG_2, BE_U8_0, BE_U8_1, BE_U8_2,
+    BE_U8_MAX, BE_U8_MIN, BE_U8_NEG_1, BE_U8_NEG_2, LE_U32_0, LE_U32_1, LE_U32_2, LE_U32_MAX, LE_U32_MIN, LE_U32_NEG_1,
+    LE_U32_NEG_2, LE_U8_0, LE_U8_1, LE_U8_2, LE_U8_MAX, LE_U8_MIN, LE_U8_NEG_1, LE_U8_NEG_2,
+};
+
+use crate::ternary::bigint::{
+    binary_representation::{
+        BinaryRepresentation, U32Repr, U8Repr, BINARY_LEN_IN_U32 as LEN_IN_U32, BINARY_LEN_IN_U8 as LEN_IN_U8,
+    },
+    endianness::{BigEndian, LittleEndian},
+    error::Error,
+    overflowing_add::OverflowingAdd,
+    u384, T242, T243, U384,
+};
+
+use bee_ternary::Btrit;
+
+use byteorder::{self, ByteOrder};
+
+use std::{
+    cmp::Ordering,
+    convert::{TryFrom, TryInto},
+    fmt,
+    marker::PhantomData,
+    ops::{Deref, DerefMut},
+};
+
+/// A big integer encoding a signed integer with 384 bits.
+///
+/// `T` is usually taken as a `[u32; 12]` or `[u8; 48]`.
+///
+/// `E` refers to the endianness of the digits in `T`. This means that in the case of `[u32; 12]`, if `E == BigEndian`,
+/// that the u32 at position i=0 is considered the most significant digit. The endianness `E` here makes no statement
+/// about the endianness of each single digit within itself (this then is dependent on the endianness of the platform
+/// this code is run on).
+///
+/// For `E == LittleEndian` the digit at the last position is considered to be the most significant.
+#[derive(Clone, Copy)]
+pub struct I384<E, T> {
+    pub(crate) inner: T,
+    _phantom: PhantomData<E>,
+}
+
+impl<E, T> Deref for I384<E, T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl<E, T> DerefMut for I384<E, T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.inner
+    }
+}
+
+impl<E: fmt::Debug, R: BinaryRepresentation, D> fmt::Debug for I384<E, R>
```

Comment:
I think the direct bounds on the `impl` are not necessary, because they are also specified within the `where` clause.

Reply:
Absolutely right